### PR TITLE
Reapprove PR after Auto Rebase Completes

### PR DESCRIPTION
**Pull request by walshmm**
## Proposed changes
In reference to issue: https://github.com/QMCPACK/qmcpack/issues/3862
Continuation of this PR: https://github.com/QMCPACK/qmcpack/pull/3994

Allows the rebase action to reapprove pr after autorebase is complete.  It approves specifically the SHA it pushes, so there shouldnt be an accidental approval.

## What type(s) of changes does this code introduce?

- Other (please describe): Github Actions Pipeline Changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

remote github runners

Tested on this PR: https://github.com/walshmm/qmcpack/pull/11
You can follow its listed history to see it was rebased and reapproved via bot before being automerged.

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: external_codes/github_actions/auto-rebase.sh
Comment (PDoakORNL) [inline external_codes/github_actions/auto-rebase.sh:130]: what happens if this push fails.  Say an attacker knows the what the hash of the rebase will be and already pushed it.  Will this action still continue and use the COMMIT_SHA?

**Review by PDoakORNL — COMMENTED** [review]
So here's the objection a cyber security guy is going to make and I don't mean to make this difficult but we'd like to argue to be able to do CI or nightlies on Summit, Frontier or Perlmutter someday.

We're hoping that Github's monitoring for SHA-1 collision attacks which are "feasible" now will prevent the following scenario, but we don't actually know what they are doing.  Certainly if it relates to the ordering of pushes with colliding hashes that's probably not good enough.

The scenario would be something like an attacker submits a PR knowing already what the hash will be once it is rebased and secures approval with the rebase action turned on. Since the attacker can determine what the hash will be before the action rebases the PR they have time to prepare a colliding commit with a payload in it.

Then they then have several possible ways to use this.  One is in my review comment.  Another is to between the actions rebase and automated approval make a forced push to remove the commit with that hash from the repo and replace it with the payload commit.

My first thought on how to prevent this would be to have this action also add a sufficient amount of truly random "salt" to a file in the commit such that predetermining the rebased PR hash would be impossible and instead the attacker would have only the time between the rebase and the approval to prepare the collision.

I'm pretty sure that force pushing to an already approved commit removes the approval.

File: external_codes/github_actions/auto-rebase.sh
Comment (walshmm) [reply external_codes/github_actions/auto-rebase.sh:130]: I can look into creating a bot account and signing commits as outlined in this article:
https://httgp.com/signing-commits-in-github-actions/

This would essentially be adding an additional unknown salt to the commit hash.

But also according to this article git is aware of and actively monitor commits for collisions:
https://github.blog/2017-03-20-sha-1-collision-detection-on-github-com/

Comment (markdewing) [reply external_codes/github_actions/auto-rebase.sh:130]: The script has "set -e" specified at the top, so the script will exit if this command exits with a non-zero error code.  And not submit the following approval code. Under what conditions would this push fail and what state would the PR be left in?

**Comment by PDoakORNL** [comment]
Test this please.

**Review by PDoakORNL — APPROVED** [review]
I think the looks good now, but we wil need an outside of ORNL reviewer.  I'm also not up to date on whether the bot account accesses are setup properly so @walshmm should weigh in on that.

**Comment by walshmm** [comment]
> I think the looks good now, but we wil need an outside of ORNL reviewer. I'm also not up to date on whether the bot account accesses are setup properly so @walshmm should weigh in on that.

I believe Paul granted it basic access, which should be enough because all the bot is doing is signing a commit.  The pushing and rebasing is done by the github actions token.  Post merge I would need to provide the secrets it uses so they can be stored and accessed via GithubActions, then probably open some test PR to give it a once over.

**Review by ye-luo — APPROVED** [review]
@PDoakORNL primarily looked at it and approved it.

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**
